File: legacy/legacy-apis/Psi.Models.ClientConfigurationModels/Account/AccountSettings.cs
using System.Collections.Generic;
using System.Linq;
using Newtonsoft.Json;

namespace Psi.Data.Models.ClientConfigurationModels.Account
{
    public class AccountSettings : SettingsBaseHelper
    {
        private Transfers _transfer;
        private ScheduledTransfers _scheduledTransfers;
        private AccountOpening _accountOpening;
        private CrossAccountSettings _crossAccount;
        private DebitCards _debitCards;
        private JointOwners _jointOwners;
        private Cards _cards;
        private Escheat _escheat;

        public AccountSettings(ISettingsBase settingsBase) : base(settingsBase)
        {
        }

        [SettingKey("Accounts.GetDebitCardsDuringAccountInquiry")]
        public bool GetDebitCardsDuringAccountInquiry
        {
            get => GetBoolValue();
            set => SetValue(value);
        }

        [SettingKey("Accounts.ShowClosedAccountsOnDashboard")]
        public bool ShowClosedAccountsOnDashboard
        {
            get => GetBoolValue();
            set => SetValue(value);
        }

        [SettingKey("Accounts.HideLoanSummaryControlWhenMemberHasNoLoans")]
        public bool HideLoanSummaryControlWhenMemberHasNoLoans
        {
            get => GetBoolValue();
            set => SetValue(value);
        }

        /// <summary>
        /// If true, show rate for shares on summary and dashboard views in HomeBanking
        /// </summary>
        [SettingKey("Accounts.ShowRateForSharesOnDashboardAndSummary")]
        public bool ShouldShowRateForSharesOnDashboardAndSummary
        {
            get => GetBoolValue();
            set => SetValue(value);
        }

        /// <summary>
        /// list of valid share categories for certificates. default: IRA
        /// </summary>
        [SettingKey("Accounts.CertificateCategoriesForRates")]
        public List<string> CertificateCategoriesForRates
        {
            get => GetListValue();
            set => SetValue(string.Join(",", value));
        }

        /// <summary>
        /// This config is currently only valid for DNA and Spectrum cores
        /// If true, nicknames will be stored in the core
        /// If false, nicknames will be stored in our HomeBankingAccountAlias table
        /// </summary>
        [SettingKey("Account.UseCoreSuffixNicknamesOnly")]
        public bool UseCoreSuffixNicknamesOnly
        {
            get => GetBoolValue();
            set => SetValue(value);
        }

        /// <summary>
        /// Comma-delimited list of account types that have checks (for check image retrieval)
        /// </summary>
        [SettingKey("X.App.HomeBanking.AccountHasChecks")]
        public List<string> AccountCategoriesWithChecks
        {
            get
            {
                var categories = GetListValue();

                return categories?.Where(x => !string.IsNullOrEmpty(x)).ToList() ?? new List<string>(); //strip out all null and string.empty values
            }
            set => SetValue(string.Join(",", value));
        }

        /// <summary>
        /// list of Account categories that are Checking accounts
        /// </summary>
        [SettingKey("X.App.HBBOL.obsCheckingAccounts")]
        public List<string> CheckingAccountCategories
        {
            get
            {
                var categories = GetListValue();

                return categories?.Where(x => !string.IsNullOrEmpty(x)).ToList() ?? new List<string>(); //strip out all null and string.empty values
            }
            set => SetValue(string.Join(",", value));
        }

        /// <summary>
        /// list of Account categories that are share accounts
        /// </summary>
        [SettingKey("X.App.HomeBanking.obsShareAccounts")]
        public List<string> ShareAccountCategories
        {
            get
            {
                var categories = GetListValue();

                return categories?.Where(x => !string.IsNullOrEmpty(x)).ToList() ?? new List<string>(); //strip out all null and string.empty values
            }
            set => SetValue(string.Join(",", value));
        }

        /// <summary>
        /// list of Account categories that are savings accounts
        /// </summary>
        [SettingKey("X.App.HomeBanking.SavingsAccounts")]
        public List<string> SavingsAccountCategories
        {
            get
            {
                var categories = GetListValue();

                return categories?.Where(x => !string.IsNullOrEmpty(x)).ToList() ?? new List<string>(); //strip out all null and string.empty values
            }
            set => SetValue(string.Join(",", value));
        }

        [SettingKey("X.App.HomeBanking.ShowMMAChecks")]
        public bool ShouldShowMmaChecks
        {
            get => GetBoolValue();
            set => SetValue(value);
        }

        /// <summary>
        /// list of Account categories that are loan home equity accounts
        /// </summary>
        [SettingKey("X.App.HomeBanking.HECheckCategory")]
        public List<string> HomeEquityCheckingAccountCategories
        {
            get
            {
                var categories = GetListValue();

                return categories?.Where(x => !string.IsNullOrEmpty(x)).ToList() ?? new List<string>(); //strip out all null and string.empty values
            }
            set => SetValue(string.Join(",", value));
        }

        public Transfers Transfers
        {
            get => _transfer ?? (_transfer = new Transfers(SettingsBase));
            set => _transfer = value;
        }

        public ScheduledTransfers ScheduledTransfers
        {
            get => _scheduledTransfers ?? (_scheduledTransfers = new ScheduledTransfers(SettingsBase));
            set => _scheduledTransfers = value;
        }

        public AccountOpening AccountOpening
        {
            get => _accountOpening ?? (_accountOpening = new AccountOpening(SettingsBase));
            set => _accountOpening = value;
        }

        public CrossAccountSettings CrossAccount
        {
            get => _crossAccount ?? (_crossAccount = new CrossAccountSettings(SettingsBase));
            set => _crossAccount = value;
        }

        public DebitCards DebitCards
        {
            get => _debitCards ?? (_debitCards = new DebitCards(SettingsBase));
            set => _debitCards = value;
        }

        public JointOwners JointOwners
        {
            get => _jointOwners ?? (_jointOwners = new JointOwners(SettingsBase));
            set => _jointOwners = value;
        }

        public Cards Cards
        {
            get => _cards ?? (_cards = new Cards(SettingsBase));
            set => _cards = value;
        }

        public Escheat Escheat
        {
            get => _escheat ?? (_escheat = new Escheat(SettingsBase));
            set => _escheat = value;
        }

        /// <summary>
        /// list of line of credit categories  
        /// </summary>
        [SettingKey("X.App.HBBOL.obsLineOfCredit")]
        public List<string> LineOfCreditCategories
        {
            get
            {
                var categories = GetListValue();

                return categories?.Where(x => !string.IsNullOrEmpty(x)).ToList() ?? new List<string>();     //strip out all null and string.empty values
            }
            set => SetValue(string.Join(",", value));
        }

        [SettingKey("Account.CoreCategoryToDisplayCategoryMappings")]
        public Dictionary<string, string> CoreCategoryToDisplayCategoryMappings
        {
            get
            {
                var value = GetValue();
                return string.IsNullOrWhiteSpace(value) ? new Dictionary<string, string>() : JsonConvert.DeserializeObject<Dictionary<string, string>>(value);
            }
            set => SetValue(value);
        }

        [SettingKey("X.App.HomeBanking.UsePreviousMicrNumber")]
        public bool UsePreviousMicrNumber
        {
            get => GetBoolValue();
            set => SetValue(value);
        }

        [SettingKey("X.App.HomeBanking.RemoteDepositAccounts")]
        public List<string> RemoteDepositAccounts
        {
            get
            {
                var categories = GetListValue();

                return categories?.Where(x => !string.IsNullOrEmpty(x)).ToList() ?? new List<string>(); //strip out all null and string.empty values
            }
            set => SetValue(string.Join(",", value));
        }

        [SettingKey("X.App.HomeBanking.AliasType1")]
        public string AliasType1
        {
            get => GetValue();
            set => SetValue(value);
        }

        [SettingKey("Account.NicknameValidationRegex")]
        public string NicknameValidationRegex
        {
            get
            {
                var value = GetValue();
                return string.IsNullOrWhiteSpace(value) ? AliasType1 : value;
            }
            set => SetValue(value);
        }

        /// <summary>
        /// If set to true, show account limits on the direct deposit information page
        /// </summary>
        [SettingKey("Account.ShowAccountLimitsOnDirectDepositPage")]
        public bool ShowAccountLimitsOnDirectDepositPage
        {
            get => GetBoolValue();
            set => SetValue(value);
        }     
    }
}

File: workspace/legacy-analyzer/output/infrastructure/models/Account/AccountSettings.ts
// Generated imports
import { CertificateCategoriesForRates } from '../CertificateCategoriesForRates';
import { AccountCategoriesWithChecks } from '../AccountCategoriesWithChecks';
import { CheckingAccountCategories } from '../CheckingAccountCategories';
import { ShareAccountCategories } from '../ShareAccountCategories';
import { SavingsAccountCategories } from '../SavingsAccountCategories';
import { HomeEquityCheckingAccountCategories } from '../HomeEquityCheckingAccountCategories';
import { Transfers } from '../Transfers';
import { ScheduledTransfers } from '../ScheduledTransfers';
import { AccountOpening } from './AccountOpening';
import { CrossAccountSettings } from '../CrossAccountSettings';
import { DebitCards } from '../DebitCards';
import { JointOwners } from '../JointOwners';
import { Cards } from '../Cards';
import { Escheat } from '../Escheat';
import { LineOfCreditCategories } from '../LineOfCreditCategories';
import { CoreCategoryToDisplayCategoryMappings } from '../CoreCategoryToDisplayCategoryMappings';
import { RemoteDepositAccounts } from '../RemoteDepositAccounts';

export interface AccountSettings {
    /** @settingKey Accounts.GetDebitCardsDuringAccountInquiry */
    getDebitCardsDuringAccountInquiry: boolean;
    /** @settingKey Accounts.ShowClosedAccountsOnDashboard */
    showClosedAccountsOnDashboard: boolean;
    /** @settingKey Accounts.HideLoanSummaryControlWhenMemberHasNoLoans */
    hideLoanSummaryControlWhenMemberHasNoLoans: boolean;
    /** @settingKey Accounts.ShowRateForSharesOnDashboardAndSummary */
    /**
     * /// <summary>
     * /// /// <summary>
     * /// /// If true, show rate for shares on summary and dashboard views in HomeBanking
     * /// /// </summary>
     * /// </summary>
     */
    shouldShowRateForSharesOnDashboardAndSummary: boolean;
    /** @settingKey Accounts.CertificateCategoriesForRates */
    /**
     * /// <summary>
     * /// /// <summary>
     * /// /// list of valid share categories for certificates. default: IRA
     * /// /// </summary>
     * /// </summary>
     */
    list: CertificateCategoriesForRates;
    /** @settingKey Account.UseCoreSuffixNicknamesOnly */
    /**
     * /// <summary>
     * /// /// <summary>
     * /// /// This config is currently only valid for DNA and Spectrum cores
     * /// /// If true, nicknames will be stored in the core
     * /// /// If false, nicknames will be stored in our HomeBankingAccountAlias table
     * /// /// </summary>
     * /// </summary>
     */
    useCoreSuffixNicknamesOnly: boolean;
    /** @settingKey X.App.HomeBanking.AccountHasChecks */
    /**
     * /// <summary>
     * /// /// <summary>
     * /// /// Comma-delimited list of account types that have checks (for check image retrieval)
     * /// /// </summary>
     * /// </summary>
     */
    list: AccountCategoriesWithChecks;
    /** @settingKey X.App.HBBOL.obsCheckingAccounts */
    /**
     * /// <summary>
     * /// /// <summary>
     * /// /// list of Account categories that are Checking accounts
     * /// /// </summary>
     * /// </summary>
     */
    list: CheckingAccountCategories;
    /** @settingKey X.App.HomeBanking.obsShareAccounts */
    /**
     * /// <summary>
     * /// /// <summary>
     * /// /// list of Account categories that are share accounts
     * /// /// </summary>
     * /// </summary>
     */
    list: ShareAccountCategories;
    /** @settingKey X.App.HomeBanking.SavingsAccounts */
    /**
     * /// <summary>
     * /// /// <summary>
     * /// /// list of Account categories that are savings accounts
     * /// /// </summary>
     * /// </summary>
     */
    list: SavingsAccountCategories;
    /** @settingKey X.App.HomeBanking.ShowMMAChecks */
    shouldShowMmaChecks: boolean;
    /** @settingKey X.App.HomeBanking.HECheckCategory */
    /**
     * /// <summary>
     * /// /// <summary>
     * /// /// list of Account categories that are loan home equity accounts
     * /// /// </summary>
     * /// </summary>
     */
    list: HomeEquityCheckingAccountCategories;
    transfers: Transfers;
    scheduledTransfers: ScheduledTransfers;
    accountOpening: AccountOpening;
    crossAccountSettings: CrossAccountSettings;
    debitCards: DebitCards;
    jointOwners: JointOwners;
    cards: Cards;
    escheat: Escheat;
    /** @settingKey X.App.HBBOL.obsLineOfCredit */
    /**
     * /// <summary>
     * /// /// <summary>
     * /// /// list of line of credit categories
     * /// /// </summary>
     * /// </summary>
     */
    list: LineOfCreditCategories;
    /** @settingKey Account.CoreCategoryToDisplayCategoryMappings */
    dictionary: CoreCategoryToDisplayCategoryMappings;
    /** @settingKey X.App.HomeBanking.UsePreviousMicrNumber */
    usePreviousMicrNumber: boolean;
    /** @settingKey X.App.HomeBanking.RemoteDepositAccounts */
    list: RemoteDepositAccounts;
    /** @settingKey X.App.HomeBanking.AliasType1 */
    aliasType1: string;
    /** @settingKey Account.NicknameValidationRegex */
    nicknameValidationRegex: string;
    /** @settingKey Account.ShowAccountLimitsOnDirectDepositPage */
    /**
     * /// <summary>
     * /// /// <summary>
     * /// /// If set to true, show account limits on the direct deposit information page
     * /// /// </summary>
     * /// </summary>
     */
    showAccountLimitsOnDirectDepositPage: boolean;
}
